File: src/test/java/com/truveta/opentoken/attributes/person/SexAttributeTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SexAttributeTest {
    private SexAttribute sexAttribute;

    @BeforeEach
    void setUp() {
        sexAttribute = new SexAttribute();
    }

    @Test
    void getName_ShouldReturnSex() {
        assertEquals("Sex", sexAttribute.getName());
    }

    @Test
    void getAliases_ShouldReturnSexAndGender() {
        String[] expectedAliases = { "Sex", "Gender" };
        assertArrayEquals(expectedAliases, sexAttribute.getAliases());
    }

    @Test
    void normalize_ShouldReturnMaleForMInput() {
        assertEquals("Male", sexAttribute.normalize("M"));
        assertEquals("Male", sexAttribute.normalize("Male"));
        assertEquals("Male", sexAttribute.normalize("m"));
        assertEquals("Male", sexAttribute.normalize("male"));
    }

    @Test
    void normalize_ShouldReturnFemaleForFInput() {
        assertEquals("Female", sexAttribute.normalize("F"));
        assertEquals("Female", sexAttribute.normalize("Female"));
        assertEquals("Female", sexAttribute.normalize("f"));
        assertEquals("Female", sexAttribute.normalize("female"));
    }

    @Test
    void normalize_ShouldReturnNullForInvalidInput() {
        assertNull(sexAttribute.normalize("X"), "Invalid value should return null");
        assertNull(sexAttribute.normalize("Other"), "Invalid value should return null");
    }

    @Test
    void validate_ShouldReturnTrueForValidValues() {
        assertTrue(sexAttribute.validate("M"));
        assertTrue(sexAttribute.validate("F"));
        assertTrue(sexAttribute.validate("Male"));
        assertTrue(sexAttribute.validate("Female"));
    }

    @Test
    void validate_ShouldReturnFalseForInvalidValues() {
        assertFalse(sexAttribute.validate("X"), "Invalid value should not be allowed");
        assertFalse(sexAttribute.validate("Other"), "Invalid value should not be allowed");
        assertFalse(sexAttribute.validate(""), "Empty value should not be allowed");
        assertFalse(sexAttribute.validate(null), "Null value should not be allowed");
    }
}

File: lib/nodejs/opentoken/test/attributes/person/SexAttribute.test.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { SexAttribute } from '../../../src/attributes/person/SexAttribute';

describe('SexAttribute', () => {
  let sexAttribute: SexAttribute;

  beforeEach(() => {
    sexAttribute = new SexAttribute();
  });

  test('getName should return Sex', () => {
    expect(sexAttribute.getName()).toBe('Sex');
  });

  test('getAliases should return Sex and Gender', () => {
    const expectedAliases = ['Sex', 'Gender'];
    const actualAliases = sexAttribute.getAliases();
    expect(actualAliases).toEqual(expectedAliases);
  });

  test('normalize should convert male variations to MALE', () => {
    expect(sexAttribute.normalize('M')).toBe('MALE');
    expect(sexAttribute.normalize('m')).toBe('MALE');
    expect(sexAttribute.normalize('Male')).toBe('MALE');
    expect(sexAttribute.normalize('MALE')).toBe('MALE');
    expect(sexAttribute.normalize('male')).toBe('MALE');
  });

  test('normalize should convert female variations to FEMALE', () => {
    expect(sexAttribute.normalize('F')).toBe('FEMALE');
    expect(sexAttribute.normalize('f')).toBe('FEMALE');
    expect(sexAttribute.normalize('Female')).toBe('FEMALE');
    expect(sexAttribute.normalize('FEMALE')).toBe('FEMALE');
    expect(sexAttribute.normalize('female')).toBe('FEMALE');
  });

  test('validate should return true for male variations', () => {
    expect(sexAttribute.validate('M')).toBe(true);
    expect(sexAttribute.validate('Male')).toBe(true);
    expect(sexAttribute.validate('MALE')).toBe(true);
  });

  test('validate should return true for female variations', () => {
    expect(sexAttribute.validate('F')).toBe(true);
    expect(sexAttribute.validate('Female')).toBe(true);
    expect(sexAttribute.validate('FEMALE')).toBe(true);
  });

  test('validate should return false for invalid values', () => {
    expect(sexAttribute.validate('X')).toBe(false);
    expect(sexAttribute.validate('Unknown')).toBe(false);
    expect(sexAttribute.validate('Other')).toBe(false);
    expect(sexAttribute.validate('')).toBe(false);
  });
});
